io-engine/src/bdev/nexus/nexus_channel.rs

```diff
@@ -193,67 +196,69 @@ impl<'n> NexusChannel<'n> {
-        // swap them out later
-
-        let mut writers = Vec::new();
-        let mut readers = Vec::new();
-
-        // iterate over all our children which are in the healthy state
-        self.nexus()
-            .children_iter()
-            .filter(|c| c.is_healthy())
-            .for_each(|c| match (c.get_io_handle(), c.get_io_handle()) {
-                (Ok(w), Ok(r)) => {
-                    writers.push(w);
-                    readers.push(r);
-                }
-                _ => {
-                    c.set_faulted_state(FaultReason::CantOpen);
-                    error!("{self:?}: failed to get I/O handle for {c:?}");
-                }
-            });
-
-        // then add write-only children
-        if !readers.is_empty() {
-            self.nexus()
-                .children_iter()
-                .filter(|c| c.is_rebuilding())
-                .for_each(|c| match c.get_io_handle() {
-                    Ok(hdl) => {
-                        debug!(
-                            "{self:?}: connecting child device \
-                                in write-only mode: {c:?}"
-                        );
-                        writers.push(hdl);
+        match dev_ctx {
+            DrEvent::ChildRebuildBegin(dev) => {
+                if let Some(child) =
+                    self.nexus().child_by_device_name(dev.as_str())
+                {
+                    if !self.readers.is_empty() {
+                        // TODO: assert that the device is !is_writer here.
+                        match child.get_io_handle() {
+                            Ok(hdl) => {
+                                debug!("{self:?}: connecting child device in write-only mode: {child:?}");
+                                self.writers.push(hdl);
+                            }
+                            Err(e) => {
+                                child.set_faulted_state(FaultReason::CantOpen);
+                                error!("{self:?}: failed to get I/O handle for {child:?}: {e}");
+                            }
+                        }
+                        // Make sure the child is not in readers. Most likely
+                        // the child won't be in readers
+                        // already here.
+                        self.readers
+                            .retain(|h| &h.get_device().device_name() != dev);
                     }
-                    Err(e) => {
-                        c.set_faulted_state(FaultReason::CantOpen);
-                        error!(
-                            "{self:?}: failed to get I/O handle \
-                                for {c:?}: {e}"
-                        );
+                }
+            }
+            DrEvent::ChildRebuildEnd(dev) => {
+                if let Some(child) =
+                    self.nexus().child_by_device_name(dev.as_str())
+                {
+                    if child.is_healthy() {
```

Comment:
Do we need to check if it's healthy here?

Reply:
Then shouldn't we remove the device from readers and writers in case the child is not healthy?

Reply:
In case child is not healthy as a result of rebuild fail/stop, the unplug workflow will kick in anyway so we'll be back here again as ChildUnplug which will do the needful.

Reply:
Isn't this different behaviour from what happens today, when rebuild fails the reconfigure removes the child from writers straight away?

Reply:
straight away in this case means via on_rebuild_update() callback from rebuild backend, but by that time unplug is much ahead and has already done the reconfigure.

Reply:
unplug IIRC is based on a callback from SPDK, so it may happen before or after depending on spdk and timing.
We should remove it on reconfigure and keep same behaviour.